### PR TITLE
tests: show unminified stacktraces for Errors

### PR DESCRIPTION
**Pull request by alxndrsn**
Maps stacktraces of Errors thrown in mocha tests run against `pouchdb.min.js` in browser tests when run with playwright.  

The mapped stacktraces are printed to stdout.

The original stackrace is still returned as part of the test failure.

### Example: https://github.com/alxndrsn/pouchdb/actions/runs/7150748499/job/19474399720

```
  test.basics.js-http
     [Create a pouch without new keyword] Minified error stacktrace re-mapped to:
        Error: Missing/invalid DB name
          _setup@pouchdb.js:5634:13
          call@pouchdb.js:5677:34
          apply@pouchdb.js:5572:10
          n@pouchdb.js:5570:14
          <unknown>@test.basics.js:20:16
          startTests@webrunner.js:82:11
          EventListener.handleEvent*@webrunner.js:90:14
          promise callback*@webrunner.js:85:27
          <unknown>@webrunner.js:94:3
    1) Create a pouch without new keyword
     [Create a pouch without new keyword] Minified error stacktrace re-mapped to:
        Error: Missing/invalid DB name
          _setup@pouchdb.js:5634:13
          call@pouchdb.js:5677:34
          apply@pouchdb.js:5572:10
          n@pouchdb.js:5570:14
          <unknown>@test.basics.js:20:16
          startTests@webrunner.js:82:11
          EventListener.handleEvent*@webrunner.js:90:14
          promise callback*@webrunner.js:85:27
          <unknown>@webrunner.js:94:3

  test.basics.js-local
     [Create a pouch without new keyword] Minified error stacktrace re-mapped to:
        Error: Missing/invalid DB name
          _setup@pouchdb.js:5634:13
          call@pouchdb.js:5677:34
          apply@pouchdb.js:5572:10
          n@pouchdb.js:5570:14
          <unknown>@test.basics.js:20:16
    2) Create a pouch without new keyword
     [Create a pouch without new keyword] Minified error stacktrace re-mapped to:
        Error: Missing/invalid DB name
          _setup@pouchdb.js:5634:13
          call@pouchdb.js:5677:34
          apply@pouchdb.js:5572:10
          n@pouchdb.js:5570:14
          <unknown>@test.basics.js:20:16


  0 passing (9ms)
  2 failing

  1) test.basics.js-http Create a pouch without new keyword:
     Missing/invalid DB name
  _setup@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:62197
  [27]</</</We<@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:63646
  n@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:61730
  n@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:61716
  @test.basics.js:20:16
  startTests@webrunner.js:82:11
  EventListener.handleEvent*@webrunner.js:90:14
  promise callback*@webrunner.js:85:27
  @webrunner.js:94:3
  

  2) test.basics.js-local Create a pouch without new keyword:
     Missing/invalid DB name
  _setup@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:62197
  [27]</</</We<@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:63646
  n@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:61730
  n@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:61716
```

Mapped stacktraces are printed twice as `mocha` fires the `fail` event twice for each test.

### CODE REVIEW

File: bin/test-browser.js [hunk 1]
```diff
@@ -87,6 +89,34 @@ class RemoteRunner {
 
       handlers[event.name].forEach(function (handler) {
         handler(obj, event.err);
+
+        if (event.err && stackConsumer) {
+          let stackMapped;
+          const mappedStack = require('stacktrace-parser')
```
Comment (lucidNTR) [inline bin/test-browser.js:95]: can we move require statements to top of file? this makes future conversion to esm much less painful and simpler to see what is going on in a module...

File: bin/test-browser.js [hunk 2]
```diff
@@ -139,6 +169,14 @@ function BenchmarkJsonReporter(runner) {
 }
 
 async function startTest() {
+  if (qs.src === '../../packages/node_modules/pouchdb/dist/pouchdb.min.js') {
+    const mapPath = './packages/node_modules/pouchdb/dist/pouchdb.min.js.map';
+    const { readFileSync } = require('fs');
+    const rawMap = readFileSync(mapPath, { encoding:'utf8' });
+    const jsonMap = JSON.parse(rawMap);
+    const { SourceMapConsumer } = require('source-map');
```
Comment (lucidNTR) [inline bin/test-browser.js:177]: see other comment about require

File: bin/build-utils.js [hunk 3]
```diff
@@ -29,8 +29,11 @@ function writeFile(filename, contents) {
 }
 
 function doUglify(pkgName, code, prepend, fileOut) {
-  var miniCode = prepend + terser.minify(code, { output: { ascii_only: true }}).code;
-  return writeFile(addPath(pkgName, fileOut), miniCode);
+  const filename = fileOut.replace(/.*\//, '');
```
Comment (lucidNTR) [inline bin/build-utils.js:32]: one of the features of source maps is being able to click the file and line number in an ide and directly open the right file and cursor position, but this requires the relative path prefixed to the filenames, probably (not sure) also a space between the prefix and filename likse so  _setup@  project-folder/filename.js

**Review by lucidNTR — CHANGES_REQUESTED** [review]
added a few minor comments. Do i understand currectly that you picked those thow packages for sourcemap handling over something like package 'source-map-support' in that this allows us to print both the translated output as well as the original error with minified references?

**Comment by lucidNTR** [comment]
excited to get this ❤️

**Comment by alxndrsn** [comment]
> Do i understand currectly that you picked those thow packages for sourcemap handling over something like package 'source-map-support' in that this allows us to print both the translated output as well as the original error with minified references?

I don't have previous experience with sourcemap handling, so the choice is somewhat arbitrary.  I do think it's important to maintain the original stacktrace to make it clear which code is actually being run.

`source-map-support` only mentions supporting Chrome in its browser documentation.  We need to support Firefox & Webkit as well.

File: bin/build-utils.js (hunk 3 above)
Comment (alxndrsn) [reply bin/build-utils.js:32]: I don't have an IDE to test this on, but the current format matches what Firefox generates.  Happy to change to other formats in this PR or in future if helpful.

Comment (alxndrsn) [reply bin/build-utils.js:32]: I'm guessing following the NodeJS format would be most compatible: https://nodejs.org/docs/latest/api/errors.html#errorstack

Comment (lucidNTR) [reply bin/build-utils.js:32]: i am not sure how the nodejs error stack format would work in the the terminal output unless i missed a section in the docu. i would not overcomplicate it and just checked vscode which seems to be the vast majority these days and it works as long as there is a space after the @ and the path is relative to the repo. eg. :

_setup@ src/main.js:3:2


this also seem the most human readable to me, what do you think?

**Comment by alxndrsn** [comment]
Updated stacktraces.  Example from https://github.com/alxndrsn/pouchdb/actions/runs/7407502604/job/20153926881:

```
  test.basics.js-local
     [Create a pouch without new keyword] Minified error stacktrace mapped to:
        Error: Missing/invalid DB name
          at _setup (pouchdb.js:5634:13)
          at call (pouchdb.js:5677:34)
          at apply (pouchdb.js:5572:10)
          at n (pouchdb.js:5570:14)
          at <unknown> (test.basics.js:20:16)
    2) Create a pouch without new keyword
     [Create a pouch without new keyword] Minified error stacktrace mapped to:
        Error: Missing/invalid DB name
          at _setup (pouchdb.js:5634:13)
          at call (pouchdb.js:5677:34)
          at apply (pouchdb.js:5572:10)
          at n (pouchdb.js:5570:14)
          at <unknown> (test.basics.js:20:16)
```

**Comment by alxndrsn** [comment]
@lucidNTR will the file path shown work (`pouchdb.js`)?  Or should it be `packages/node_modules/pouchdb/dist/pouchdb.js`?

**Comment by lucidNTR** [comment]
@alxndrsn   in the linked run it looks like this:

_setup@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:62197


this allready is much beetter as i can see the path and the link is parseable to the vscode and github out put, however as the server is not running anymore by the time the test error is usually seen its not possible to use the "click to source" feature. if you want to do it perfect you would have to remove the "http://127.0.0.1:8000/" as you said and also make sure there is a space between _setup@ and the /package/.... part

**Comment by alxndrsn** [comment]
> @alxndrsn in the linked run it looks like this:
> 
> _setup@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:62197
> 
> this allready is much better (in the original version it was just shown as [_setup@pouchdb.min.js](mailto:_setup@pouchdb.min.js):7:62197) 

I don't think there have been any changes made to `_setup@http://127.0.0.1:8000/packages/node_modules/pouchdb/dist/pouchdb.min.js:7:62197` - that looks like standard output from chromium.

**Comment by lucidNTR** [comment]
@alxndrsn  hmm maybe i missread something. what was the change in your comment here then? https://github.com/pouchdb/pouchdb/pull/8841#issuecomment-1876773226

**Comment by alxndrsn** [comment]
@lucidNTR now using NodeJS format, which is "at $fnName ($file:$line:$column)", and the mapped link should not point to the `.min.js` file.

**Comment by garethbowen** [comment]
@lucidNTR Would you mind reviewing this again?

**Review by lucidNTR — APPROVED** [review]
@alxndrsn  @garethbowen added approval, but would really like to have the inlined require  calls move to the top of file if not too much hassle. (readability, more inline with esm style expectations in which inline would always be async dynamic imports)

**Pull request merged**